### ISSUE TITLE
[easy] Clarify Lwt.finalize documentation

### ISSUE THREAD

**Issue by aantron** [ocsigen/lwt#442]
From https://discuss.ocaml.org/t/lwt-finalize-and-failure-in-the-cleanup-code/525:

> The documentation for Lwt.finalize states:
>
> > `finalize f g` returns the same result as `f ()` whether it fails or not. In both cases, `g ()` is executed after `f`.
>
> Yet the following example demonstrates that this is not the case:
>
> ```ocaml
> Lwt.finalize
>   (fun () -> Lwt.fail (Failure "1"))
>   (fun () -> Lwt.fail (Failure "2"))
> |> state;;
> ```
>
> This gives a state with `Failure "2"`, not the `Failure "1"` so the failure comes from the execution of g. Is it a bug in the documentation?

Indeed, if `g ()` fails, that behavior controls the behavior of `finalize`, no matter the outcome of `f ()`. The documentation does not mention this, and states unconditionally that `finalize` returns the same result as `f ()`.

The documentation is found here: https://github.com/ocsigen/lwt/blob/99f556b347567f92a1b290915881692b94b060b4/src/core/lwt.mli#L164-L166

**Comment by Richard-Degenne** [comment]
What's a use-case for this function? As you stated [here](https://discuss.ocaml.org/t/lwt-finalize-and-failure-in-the-cleanup-code/525/2?u=richard-degenne), it might be interested to discuss whether it is the docs or the behavior that needs to be changed.

**Comment by aantron** [comment]
Resource cleanup, for example. Even though there is `Lwt_io.with_file`, I'm using [`Lwt_io.open_file`](https://ocsigen.org/lwt/dev/api/Lwt_io#VALopen_file) to demonstrate:

```ocaml
let%lwt file = Lwt_io.(open_file Input "foo.txt") in
Lwt.finalize
  (fun () ->
    (* read the file *))
  (fun () ->
    Lwt_io.close file)
```

In fact, this is how `with_file` is [implemented](https://github.com/ocsigen/lwt/blob/99f556b347567f92a1b290915881692b94b060b4/src/unix/lwt_io.ml#L1351-L1355).

As I understand it, in Python, if you raise inside `finally:`, that exception also overrules any exception that may have triggered the `finally:` clause to begin with. I haven't looked yet for the semantics in Java, or in other languages.

**Comment by Richard-Degenne** [comment]
Oh, I see. I just gave Java a try, and the behaviour is the same: the exception in the `finally` block hides the other one.

The behavior is documented in the [Java language specifications](https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.20.2), which is clearly not a resource intended for developers, but we can draw inspiration from it.

**Comment by aantron** [comment]
Thanks for checking.

I think we should find the best justifications for the various behaviors, choose one (or make our own :scream:), document the design decision, and make these links easy for users to find. We may not be able to change `Lwt.finalize`, if that's what we want to do, but we have opportunities to create new, better APIs coming up.

**Comment by Richard-Degenne** [comment]
Well, the current behavior is fine with me. One way to interpret it would be as follows: `Lwt.finalize` (or any `try-finally` construct, really) is to be considered as one semantic entity.

- If the `try` part fails, the `finally` part executes;
  - If it fails as well, the `try-finally` fails *because* of the `finally`'s failure.
  - If the `finally` completes/resolves, the `try-finally` still fails, but because of the `try`'s failure.

This is kinda hard of express in layman's terms, but I think you get the idea.

Wording it properly for the documentation may require some effort, though.

**Comment by aantron** [comment]
It might be relatively easy, e.g.

> `Lwt_finalize f g` runs `f ()`. Whether `f ()` returns normally or fails, `Lwt.finalize` then runs `g ()`. The whole expression returns the same result as `f ()`, unless `g ()` fails, in which case the result is the exception raised by `g ()`.

This is just a sketch. Some editing will be needed :)

It may also help to rename `g` to something more descriptive, such as `finally` or `cleanup`.

**Comment by Richard-Degenne** [comment]
We could even consider making it a labeled argument. It would make the whole thing and its usages far more expressive, I think.

```
Lwt.finalize (fun () -> Lwt.fail_with "try failed!")
  ~finally:(fun () -> Lwt.fail_with "finally failed!");;
```

I'm not sure what the implications in terms of retro-compatibility are, though.

**Comment by aantron** [comment]
I agree with the labeled argument (the same for `catch`); I routinely defined this in my projects. It won't be backwards-compatible though, except for people using the option to ignore labels (is it still the default?).

There might be a future new `Lwt.Promise` module, in which we can create a new API of our dreams. So these discussions are welcome, even if we can't really change `Lwt.finalize` itself too much.
